language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs

```diff
@@ -187,26 +188,52 @@ proceeds as follows:
 
       PLC.reverse {bool} (cons true (cons false nil))
     ~ makeKnown (Haskell.reverse (readKnown (cons true (cons false nil))))
-    ~ makeKnown (Haskell.reverse [OpaqueTerm true, OpaqueTerm false])
-    ~ makeKnown [OpaqueTerm false, OpaqueTerm true]
-    ~ cons false (cons true nil)
+    ~ makeKnown (Haskell.reverse [Opaque true, Opaque false])
+    ~ makeKnown [Opaque false, Opaque true]
+    ~ EvaluationSuccess (cons false (cons true nil))
 
-Note how we use the 'OpaqueTerm' wrapper in order to unlift a PLC term as an opaque Haskell value
+Note how we use the 'Opaque' wrapper in order to unlift a PLC term as an opaque Haskell value
 using 'readKnown' and then lift the term back using 'makeKnown' without ever inspecting the term.
 
+@Opaque term rep@ can be used for passing any @term@ through the builtin application machinery,
+not just one whose type is a bound variable. For example, you can define a new data type
+
+    data NatRep
+
+provide a 'KnownTypeAst' instance for it (mapping a @Proxy NatRep@ to the actual type of natural
+numbers in PLC) and define a (rather pointless) builtin like @idNat : nat -> nat@.
+
+It's also possible to bind a type variable of a higher-kind, say, @f :: * -> *@ and make a builtin
+with the following signature:
+
+    idFInteger : all (f :: * -> *). f integer -> f integer
+
+where the type application is handled with 'TyAppRep'
+
+It would be much better if we didn't need to define that @*Rep@ stuff at the type level just to
+demote it to the term level via a type class, but this hasn't been investigated yet. A plausible
+way would be to ditch 'KnownTypeAst' (but keep 'KnownType') and provide PLC types manually.
+But that doesn't seem to give rise to a terribly nice API.
```

Comment:
I don't know if we need this or not. The current solution feels pretty usable and I kind of like that we reflect more stuff at the type-level (like variable names/uniques). Without the flexible builtins that we had before and with the alternative CEK machine we won't be able to juggle PLC terms (like, apply one term to another etc) in the interpretation of a builtin, so this use case seems to be gone. And it doesn't seem like using PLC types in builtins is very useful.

I think I'll make an issue in JIRA and won't touch this little problem for now (I spent some time on it, but I was unable to quickly find a nice solution).